### ISSUE TITLE
Interface nil check issue in http.NewRequest

### ISSUE THREAD

**Issue by nathanjsweet** [golang/go#19429]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?
1.7.4
### What operating system and processor architecture are you using (`go env`)?
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/nathan/Development/golang"
GORACE=""
GOROOT="/usr/local/go"
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
CC="gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build075596834=/tmp/go-build -gno-record-gcc-switches"
CXX="g++"
CGO_ENABLED="1"

### What did you do?
https://play.golang.org/p/5xtRNTx6xB

We all know what I just posted happens, however it seems that the go team itself struggles with this, per:

https://github.com/golang/go/blob/9d29be468eb9092f9dea3e10d32e1f7848a55458/src/net/http/request.go#L796
Why is the nil check done in `http.NewRequest` before the type checking? It should be the other way around. I just ran into this and it's really messing me up. 

To summarize line 799 will panic if a nil *bytes.Buffer is passed to `NewRequest`.

\<rant\>
If nil checking interfaces isn't intuitive to you guys how is it supposed to be for the rest of us? This is easily my least favorite part of golang.
\</rant\>

**Comment by mvdan** [comment]
Comparing to nil can still be useful if both the type and value are nil. What is your point? Is that bit of code in net/http causing any bug?

Also, note that you can make suggestions without ranting. You're also much more likely to get constructive replies if you write constructively yourself.

**Comment by nathanjsweet** [comment]
Yes the bug is that if you pass a nil `*bytes.Buffer` type then (from the commit referenced in the link I posted) line 799 will panic. I would expect that I could pass a nil `*bytes.Buffer` without `http.NewRequest` panicking.

**Comment by bradfitz** [comment]
>  I would expect that I could pass a nil *bytes.Buffer without http.NewRequest panicking.

That is not a correct expectation, because a nil `*bytes.Buffer` is not a functional `io.Reader`, even though it implements `io.Reader`. It's just a bogus as passing an exhausted `io.Reader` to any other function accepting an `io.Reader`.

The code is correct. It's checking for a nil interface value, not a nil pointer inside a non-nil interface value.

**Comment by nathanjsweet** [comment]
>  is not a functional io.Reader

What does that even mean?

**Comment by bradfitz** [comment]
For questions about Go, see https://golang.org/wiki/Questions.
